nixos/facter: add Intel IPU6 camera support

Mic92 opened the pull request:
This adds automatic detection and configuration for Intel IPU6 cameras:

This is particularly useful for modern Intel laptops (11th-14th gen) with integrated webcams that require the intel-ipu6 drivers. Auto-configuration eliminates manual setup for these specialized devices.

Builds on PR #463457 (input peripherals).
Part of incremental upstreaming from nixos-facter-modules.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Mic92 committed:
nixos/facter: add Intel IPU6 camera support

This adds automatic detection and configuration for Intel IPU6 cameras:

This is particularly useful for modern Intel laptops (11th-14th gen) with
integrated webcams that require the intel-ipu6 drivers. Auto-configuration
eliminates manual setup for these specialized devices.

Builds on PR #463457 (input peripherals).
Part of incremental upstreaming from nixos-facter-modules.

hfxbse reviewed (commented):
Looks fine, but somehow I am unable to rebuild the system configuration when removing the external modules and [switching to the pull request branch as Nixpkgs input URL](https://github.com/hfxbse/nixos-config/blob/a6e5959269c69a2afa3d06d722aec844134ad48d/flake.nix#L5) like so

```nix
inputs = {
  nixpkgs.url = "github:Mic92/nixpkgs?ref=facter-camera";
  ...
};
```

```
building the system configuration...
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating derivation 'nixos-system-ice-skate-26.05.20251201.cb6d783'
         whose name attribute is located at /nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/pkgs/stdenv/generic/make-derivation.nix:541:13

       … while evaluating attribute 'buildCommand' of derivation 'nixos-system-ice-skate-26.05.20251201.cb6d783'
         at /nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/system/activation/top-level.nix:69:7:
           68|       passAsFile = [ "extraDependencies" ];
           69|       buildCommand = systemBuilder;
             |       ^
           70|

       … while evaluating the option `system.systemBuilderCommands':

       … while evaluating definitions from `/nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/system/boot/kernel.nix':

       … while evaluating the option `boot.initrd.systemd.storePaths."[definition 3-entry 37]".source':

       … while evaluating definitions from `/nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/system/boot/systemd/initrd.nix':

       … while evaluating the option `boot.initrd.systemd.contents."/etc/modules-load.d/nixos.conf".source':

       … while evaluating definitions from `/nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/system/boot/systemd/initrd.nix':

       … while evaluating the option `boot.initrd.systemd.contents."/etc/modules-load.d/nixos.conf".text':

       … while evaluating definitions from `/nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/system/boot/systemd/initrd.nix':

       … while evaluating the option `boot.initrd.kernelModules':

       … while evaluating definitions from `/nix/store/bmc6yp84i66la5q62siw6bdlf69mg9dy-source/nixos/modules/hardware/facter/graphics':

       … while evaluating the option `hardware.facter.detected.boot.graphics.kernelModules':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: expected a string but found a list: [ "i915" ]
```

I believe this is not caused by this PR directly? Did I misunderstand something on how to test this PR? I also reran `nixos-facter` to generate a new report to be sure but this issue persists. Other hosts configuration result in a similar error but with a different kernel module name inside the list. All config and and report file are in [my repository](https://github.com/hfxbse/nixos-config/tree/facter-upstream).

Mic92 committed:
nixos/facter: fix collectDrivers returning nested lists

The collectDrivers function was using lib.catAttrs which extracts
attribute values into a list. Since driver_modules is already a list,
this resulted in nested lists like [["i915"]] instead of ["i915"],
causing boot.initrd.kernelModules to fail with:

  error: expected a string but found a list: [ "i915" ]

Restore the original foldl' implementation from nixos-facter-modules
that properly concatenates driver_modules lists into a flat result.

Add graphics_card and monitor entries to test data to catch this
regression.

Fixes: https://github.com/NixOS/nixpkgs/pull/466808#issuecomment-2749380083

Mic92 committed:
nixos/facter: fix ipu6 kernel version warning condition

The suspension/shutdown regression affects kernel 6.16 and later, not
just 6.16 specifically. Use versionAtLeast instead of equality check.

Also fix issue reference (372 was a separate issue) and typo.

Mic92 commented on nixos/modules/hardware/facter/camera/ipu6.nix:
```diff
@@ -0,0 +1,115 @@
+      A regression bug can occur when combining ipu6 with kernel version 6.16.
+      This will most likely prevent your system from properely suspending or shutting down.
+      For more information see: https://github.com/intel/ipu6-drivers/issues/372
+    '';
```
Updated.
